java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/SlidingWindowUtils.java

```diff
@@ -0,0 +1,186 @@
+                                    .as(aggregationFieldDescriptor.outFieldName));
+        }
+        return table;
+    }
+
+    private static List<DataTypes.Field> getTableFields(
+            String[] keyFieldNames,
+            String rowTimeFieldName,
+            AggregationFieldsDescriptor aggregationFieldsDescriptor,
+            ResolvedSchema resolvedSchema) {
+        List<DataTypes.Field> keyFields =
+                Arrays.stream(keyFieldNames)
+                        .map(
+                                fieldName ->
+                                        DataTypes.FIELD(
+                                                fieldName, getDataType(resolvedSchema, fieldName)))
+                        .collect(Collectors.toList());
+        List<DataTypes.Field> aggFieldDataTypes =
+                aggregationFieldsDescriptor.getAggFieldDescriptors().stream()
+                        .map(d -> DataTypes.FIELD(d.outFieldName, d.aggFunc.getResultDatatype()))
+                        .collect(Collectors.toList());
+        final List<DataTypes.Field> fields = new LinkedList<>();
+        fields.addAll(keyFields);
+        fields.addAll(aggFieldDataTypes);
+        fields.add(
+                DataTypes.FIELD(rowTimeFieldName, getDataType(resolvedSchema, rowTimeFieldName)));
+        return fields;
+    }
+
+    private static DataType getDataType(ResolvedSchema resolvedSchema, String fieldName) {
+        return resolvedSchema
+                .getColumn(fieldName)
+                .orElseThrow(
+                        () ->
+                                new RuntimeException(
+                                        String.format("Cannot find column %s.", fieldName)))
+                .getDataType();
+    }
+
+    private static Schema getResultTableSchema(
+            ResolvedSchema resolvedSchema,
+            AggregationFieldsDescriptor descriptor,
+            String rowTimeFieldName,
+            String[] keyFieldNames) {
+        final Schema.Builder builder = Schema.newBuilder();
+
+        for (String keyFieldName : keyFieldNames) {
+            builder.column(keyFieldName, getDataType(resolvedSchema, keyFieldName).notNull());
+        }
+
+        if (keyFieldNames.length > 0) {
+            builder.primaryKey(keyFieldNames);
+        }
+
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggregationFieldDescriptor :
+                descriptor.getAggFieldDescriptors()) {
+            builder.column(
+                    aggregationFieldDescriptor.outFieldName,
+                    aggregationFieldDescriptor.aggFunc.getResultDatatype());
+        }
+
+        builder.column(rowTimeFieldName, getDataType(resolvedSchema, rowTimeFieldName));
+
+        // Records are ordered by row time after sliding window.
+        builder.watermark(
+                rowTimeFieldName,
+                String.format("`%s` - INTERVAL '0.001' SECONDS", rowTimeFieldName));
```

Comment:
Since we have chosen to use `window_time` emitted by the tumble window function, and window_time = window_end - 1 ms, do we still need to subtract 1 ms here?

Reply:
Unfortunately, I think we still need to subtract 1 ms here. The main reason is not that the orders of the rows but that rows can have the same timestamp. [reference](https://stackoverflow.com/questions/54155126/does-apache-flink-support-multiple-events-with-the-same-timestamp)

> A Watermark is an assertion that all events that follow will have timestamps greater than the watermark, so this does mean that you cannot safely emit a Watermark t for a stream element at time t, unless the timestamps in the stream are strictly monotonically increasing -- which is not the case if there are multiple events with the same timestamp. This is why the AscendingTimestampExtractor produces watermarks equal to currentTimestamp - 1, and you should do the same.